### PR TITLE
[llvm][AArch64] Autoupgrade function attributes from Module attributes.

### PR DESCRIPTION
**Pull request by DanielKristofKiss**
`sign-return-address` and similar module attributes should be propagated to
the function level before got merged because module flags may contradict and
this information is not recoverable.
Generated code will match with the normal linking flow.

### CODE REVIEW

**Comment by DanielKristofKiss** [comment]
Updates with comments in #80642

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 1]
```diff
@@ -5155,7 +5155,51 @@ struct StrictFPUpgradeVisitor : public InstVisitor<StrictFPUpgradeVisitor> {
...
+                              StringRef ModAttrName,
+                              std::pair<StringRef, StringRef> Values) {
+  Module *M = F.getParent();
+  assert(M && "Missing module");
+  if (F.hasFnAttribute(FnAttrName))
+    return;
+  if (isModuleAttributeSet(M, ModAttrName))
+    F.addFnAttr(FnAttrName, Values.first);
+  else
+    F.addFnAttr(FnAttrName, Values.second);
```
Comment (nickdesaulniers) [inline llvm/lib/IR/AutoUpgrade.cpp:5181]: Does a ternary statement make this more concise?

Comment (DanielKristofKiss) [reply llvm/lib/IR/AutoUpgrade.cpp]: agree

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 2]
```diff
@@ -5155,7 +5155,51 @@ struct StrictFPUpgradeVisitor : public InstVisitor<StrictFPUpgradeVisitor> {
...
 } // namespace
 
-void llvm::UpgradeFunctionAttributes(Function &F) {
+// Check if the module attribute is present and not zero.
+static bool isModuleAttributeSet(const Module *M, const StringRef &ModAttr) {
+  if (const auto *Attr =
+          mdconst::extract_or_null<ConstantInt>(M->getModuleFlag(ModAttr)))
+    if (Attr->getZExtValue())
+      return true;
+  return false;
```
[the remark this answers is not in the document]
Comment (DanielKristofKiss) [reply llvm/lib/IR/AutoUpgrade.cpp]: applied

**Comment by nickdesaulniers** [comment]
Patch LGTM but I wanted to follow up on [the point about clang code gen](https://github.com/llvm/llvm-project/pull/80642#issuecomment-1942098974) from the other thread.

>> Also, I'm kind of surprised you didn't have to change anything in clang codegen of LLVM IR. Shouldn't clang now be generating fn attrs rather than module attrs?
> Clang can't really do much here as we add more functions to the module during later passes ( c++ exception handler, sanitisers especially). So the right thing is to derive the attributes from the module flag.

Right, I do recall issues with functions that have been synthesized by the compiler lacking certain function attributes...I thought we had another way of solving those (that was not AutoUpgrade).  I think that's the point of `llvm::Function::createWithDefaultAttr`. Can you use that instead?

> Autoupgrade now do this at the last moment before the functions merged into the common module.

Does auto upgrade run as a pass in the default pipelines?  I thought it was just for helping folks move from one version of IR to another (for .bc or .ll files).

**Comment by DanielKristofKiss** [comment]
> Patch LGTM but I wanted to follow up on [the point about clang code gen](https://github.com/llvm/llvm-project/pull/80642#issuecomment-1942098974) from the other thread.
> 
> > > Also, I'm kind of surprised you didn't have to change anything in clang codegen of LLVM IR. Shouldn't clang now be generating fn attrs rather than module attrs?
> > > Clang can't really do much here as we add more functions to the module during later passes ( c++ exception handler, sanitisers especially). So the right thing is to derive the attributes from the module flag.
> 
> Right, I do recall issues with functions that have been synthesized by the compiler lacking certain function attributes...I thought we had another way of solving those (that was not AutoUpgrade). I think that's the point of `llvm::Function::createWithDefaultAttr`. Can you use that instead?

backend handles the model vs function attribute e.g. llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp::GetSignReturnAddress  

So far the module attribute is the contract between the fronted/backend and I know e.g. Rust relies on that too. 

Clang also just sets the module flags today if the flag is not function attribute, but I have a patch for that too but I need to add the `llvm::Function::createWithDefaultAttr` to it.

> > Autoupgrade now do this at the last moment before the functions merged into the common module.
> 
> Does auto upgrade run as a pass in the default pipelines? I thought it was just for helping folks move from one version of IR to another (for .bc or .ll files).

That is the change that I added here, [llvm/lib/Linker/IRMover.cpp](https://github.com/llvm/llvm-project/pull/80640/files#diff-4501da671276f6ba33e0701455975409a1f9f303c57a2e044a632e3d30a9947fR1609)
run the upgrade whenever 2 module got merged, so before we loose the module flags.

**Review by nickdesaulniers — APPROVED** [review]

**Comment by nickdesaulniers** [comment]
> backend handles the model vs function attribute e.g. llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp::GetSignReturnAddress

Ah! Ok I was missing that either are already supported; I was thinking that function attributes would be new, but it looks like there is pre-existing support as you cite.

> Clang also just sets the module flags today if the flag is not function attribute, but I have a patch for that too but I need to add the llvm::Function::createWithDefaultAttr to it.

You can CC for review on that, too, when it's ready.

**Comment by nikic** [comment]
This also causes an unreasonably large compile-time regression for what it does: https://llvm-compile-time-tracker.com/compare.php?from=354401f8d3dc08ed41895d03a12a122e9cc0482c&to=531e8c26b3f2626e7f1a997e0e8b61d67d10aded&stat=instructions:u Surely upgrading attributes should not take 0.5% compile-time during the clang thin link, especially on a target where these attributes do not even exist?

This logic should probably not be inside UpgradeFunctionAttributes() -- you'll want to structure this the other way around and first determine which module attributes (if any) have to be applied, and then apply them to all functions.

**Comment by nikic** [comment]
Also, should this upgrade *really* run during normal bitcode reading? It makes little sense to me that *current* bitcode will no longer round-trip between llvm-as / llvm-dis after this change. Should the change here be limited to IRMover only?

**Comment by DanielKristofKiss** [comment]
Original change is reverted.
I refactored the patch to run only from the IRMover only and should be faster too when it doesn't need to do anything.
I'd reopen this PR but seem I can't...

**Comment by DanielKristofKiss** [comment]
#82763 is opened as I can't reopen this one.

**Pull request merged**